Userland/Services/SystemServer/DeviceEventLoop.cpp

```diff
@@ -0,0 +1,252 @@
+}
+
+static DeviceEventLoop::device_node_match const* device_node_family_to_match_type(bool block_device, MajorNumber major_number, MinorNumber minor_number)
+{
+    for (auto& matcher : s_matchers) {
+        if (matcher.major_number == major_number
+            && block_device == matcher.block_device
+            && is_in_minor_number_range(matcher, minor_number))
+            return &matcher;
+    }
+    return nullptr;
+}
+
+static bool is_in_family_minor_number_range(DeviceNodeFamily const& family, MinorNumber minor_number)
+{
+    return family.base_minor_number() <= minor_number && static_cast<MinorNumber>(family.base_minor_number().value() + family.devices_symbol_suffix_allocation_map().size()) >= minor_number;
+}
+
+DeviceNodeFamily* DeviceEventLoop::find_device_node_family(bool block_device, MajorNumber major_number, MinorNumber minor_number) const
+{
+    DeviceNodeFamily::Type type = block_device ? DeviceNodeFamily::Type::CharacterDevice : DeviceNodeFamily::Type::CharacterDevice;
+    for (auto& family : m_device_node_families) {
+        if (family->major_number() == major_number && type == family->type() && is_in_family_minor_number_range(*family, minor_number))
+            return family.ptr();
+    }
+    return nullptr;
+}
+
+ErrorOr<DeviceNodeFamily*> DeviceEventLoop::find_or_register_new_device_node_family(device_node_match const& match, bool block_device, MajorNumber major_number, MinorNumber minor_number)
+{
+    if (auto* family = find_device_node_family(block_device, major_number, minor_number); !!family)
+        return family;
+    unsigned allocation_map_size = 1024;
+    if (match.minor_number_allocation_type == MinorNumberAllocationType::SequentialLimited)
+        allocation_map_size = match.minor_number_range_size;
+    auto bitmap = TRY(Bitmap::create(allocation_map_size, false));
+    auto node = TRY(adopt_nonnull_own_or_enomem(new (nothrow) DeviceNodeFamily(move(bitmap),
+        match.family_type_literal,
+        block_device ? DeviceNodeFamily::Type::BlockDevice : DeviceNodeFamily::Type::CharacterDevice,
+        major_number,
+        minor_number)));
+    TRY(m_device_node_families.try_append(move(node)));
+
+    return m_device_node_families.last().ptr();
+}
+
+inline char offset_character_with_number(char base_char, size_t offset)
+{
+    char offsetted_char = base_char;
+    VERIFY(static_cast<size_t>(offsetted_char) + static_cast<size_t>(offset) <= static_cast<size_t>('z'));
+    offsetted_char += offset;
+    return offsetted_char;
+}
+
+static ErrorOr<String> build_suffix_with_letters(size_t allocation_index)
+{
+    auto base_string = TRY(String::from_utf8(""sv));
+    do {
+        auto current_char = offset_character_with_number('a', allocation_index % 26);
+        base_string = TRY(String::formatted("{:c}{}", current_char, base_string));
+        allocation_index = allocation_index / 26;
+    } while (allocation_index > 0);
+    return base_string;
+}
```

Comment:
nit: This could be written a bit terser

```suggestion
static ErrorOr<String> build_suffix_with_letters(size_t allocation_index)
{
    auto base_string = TRY(String::from_utf8(""sv));
    do {
        base_string = TRY(String::formatted("{:c}{}", 'a' + (allocation_index % 26), base_string));
        allocation_index = allocation_index / 26;
    } while (allocation_index > 0);
    return base_string;
}
```